test/extensions/filters/http/ext_proc/client_test.cc

```diff
@@ -0,0 +1,172 @@
+#include "envoy/config/core/v3/grpc_service.pb.h"
+
+#include "common/grpc/common.h"
+#include "common/http/header_map_impl.h"
+
+#include "extensions/filters/http/ext_proc/client_impl.h"
+
+#include "test/mocks/grpc/mocks.h"
+#include "test/mocks/stats/mocks.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using envoy::service::ext_proc::v3alpha::ProcessingRequest;
+using envoy::service::ext_proc::v3alpha::ProcessingResponse;
+
+using testing::Invoke;
+using testing::Unused;
+
+using namespace std::chrono_literals;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ExternalProcessing {
+namespace {
+
+class ExtProcStreamTest : public testing::Test, public ExternalProcessorCallbacks {
+public:
+  ~ExtProcStreamTest() override = default;
+
+protected:
+  void SetUp() override {
+    envoy::config::core::v3::GrpcService service;
+    auto grpc = service.mutable_envoy_grpc();
+    grpc->set_cluster_name("test");
+
+    EXPECT_CALL(client_manager_, factoryForGrpcService(_, _, _))
+        .WillOnce(Invoke(this, &ExtProcStreamTest::doFactory));
+
+    client_ = std::make_unique<ExternalProcessorClientImpl>(client_manager_, service, stats_store_);
+  }
+
+  Grpc::AsyncClientFactoryPtr doFactory(Unused, Unused, Unused) {
+    auto factory = std::make_unique<Grpc::MockAsyncClientFactory>();
+    EXPECT_CALL(*factory, create()).WillOnce(Invoke(this, &ExtProcStreamTest::doCreate));
+    return factory;
+  }
+
+  Grpc::RawAsyncClientPtr doCreate() {
+    auto async_client = std::make_unique<Grpc::MockAsyncClient>();
+    EXPECT_CALL(*async_client,
+                startRaw("envoy.service.ext_proc.v3alpha.ExternalProcessor", "Process", _, _))
+        .WillOnce(Invoke(this, &ExtProcStreamTest::doStartRaw));
+    return async_client;
+  }
+
+  Grpc::RawAsyncStream* doStartRaw(Unused, Unused, Grpc::RawAsyncStreamCallbacks& callbacks,
+                                   const Http::AsyncClient::StreamOptions& options) {
+    EXPECT_GT(*options.timeout, 0ms);
+    stream_callbacks_ = &callbacks;
+    return &stream_;
+  }
+
+  // ExternalProcessorCallbacks
+  void onReceiveMessage(std::unique_ptr<ProcessingResponse>&& response) override {
+    last_response_ = std::move(response);
+  }
+
+  void onGrpcError(Grpc::Status::GrpcStatus status) override { grpc_status_ = status; }
+
+  void onGrpcClose() override { grpc_closed_ = true; }
+
+  std::unique_ptr<ProcessingResponse> last_response_;
+  Grpc::Status::GrpcStatus grpc_status_ = Grpc::Status::WellKnownGrpcStatus::Ok;
+  bool grpc_closed_ = false;
+
+  std::unique_ptr<ExternalProcessorClientImpl> client_;
```

Comment:
Nit: we usually setup an alias with `using ExternalProcessClientImplPtr = std::unique..` for these smart pointer types in Envoy.

Reply:
Ack.